src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
Hmm, then it seems the call to improve performance on the [Benchmarking page](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/benchmarks.md) in the docs is a bit misguiding. Cause if I decided to try improving a rule's speed and found that using a tool slows it down that much, I woud definitely suggest ditching it. So are the lines [about writing a rule](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#write-the-rule), wich suggest considering `postcss-selctor-parser`, not require to do so. @davidtheclark could you please clarify this?

@evilebottnawi please don't think that I'm questioning your skills or experience. I'm just really confused here.